### PR TITLE
PSIMA-61,63,64 : Adding PGLIB formulation

### LINKED ISSUE #195: review UC formulations
Review the UC formulations with respect to 

https://github.com/power-grid-lib/pglib-uc/blob/master/MODEL.pdf

### PR DESCRIPTION
**Pull request by sourabhdalvi**
closes #195

### CODE REVIEW

**Comment by jd-lara** [comment]
@sourabhdalvi I changed the base of this PR to dev0.6 to avoid conflicts with the changes from @daniel-thom.

**Comment by sourabhdalvi** [comment]
Most changes are in, still adding doc strings

**Comment by jd-lara** [comment]
Remember to add a man page in the docs please.

File: src/devices_models/devices/common/range_constraint.jl [hunk 1]
```diff
@@ -107,7 +107,7 @@ function device_semicontinuousrange(
 )
     time_steps = model_time_steps(psi_container)
     varcts = get_variable(psi_container, inputs.variable_name)
-    varbin = get_variable(psi_container, inputs.bin_variable_name)
+    varbin = get_variable(psi_container, inputs.bin_variable_name[1])
```
Comment (daniel-thom) [inline src/devices_models/devices/common/range_constraint.jl:110]: Can you add asserts for the expected sizes of inputs.bin_variable_name in all functions?

File: src/devices_models/devices/common/device_range_constraints.jl [hunk 2]
```diff
@@ -107,23 +113,29 @@ function device_range_constraints!(
...
             if isnothing(limits)
                 limits = (min = 0.0, max = 0.0)
                 @warn "Range constraint limits of $T $dev_name are nothing. Set to" limits
             end
-            constraint_info = DeviceRangeConstraintInfo(dev_name, limits)
+            if constraint_struct == DeviceRangeConstraintInfo
+                constraint_info = DeviceRangeConstraintInfo(dev_name, limits)
+            elseif constraint_struct == DeviceMultiStartRangeConstraintsInfo
+                lag_limits = rc.lag_limits_func(dev)
+                constraint_info = DeviceMultiStartRangeConstraintsInfo(dev_name, limits, lag_limits)
```
Comment (daniel-thom) [inline src/devices_models/devices/common/device_range_constraints.jl]: Can you add `else error("missing implementation for $constraint_struct")`?

File: src/devices_models/devices/common/device_range_constraints.jl [hunk 3]
```diff
@@ -1,24 +1,30 @@
 struct RangeConstraintInputs
     constraint_name::String
     variable_name::String
-    bin_variable_name::Union{Nothing, String}
+    bin_variable_name::Vector{String}
```
Comment (daniel-thom) [inline src/devices_models/devices/common/device_range_constraints.jl:4]: I think this variable name should now be plural.

File: src/devices_models/devices/thermal_generation.jl [hunk 4]
```diff
@@ -187,6 +270,97 @@ function custom_active_power_constraints!(
...
+    feedforward::Union{Nothing, AbstractAffectFeedForward},
+) where {S <: PM.AbstractPowerModel}
+
+    time_steps = model_time_steps(psi_container)
+    resolution = model_resolution(psi_container)
+    key_power = ICKey(DevicePower, PSY.ThermalMultiStart)
+    key_status = ICKey(DeviceStatus, PSY.ThermalMultiStart)
+    initial_conditions_power = get_initial_conditions(psi_container, key_power)
+    initial_conditions_status = get_initial_conditions(psi_container, key_status)
+    ini_conds = _get_data_for_range_ic(initial_conditions_power, initial_conditions_status)
```
Comment (daniel-thom) [inline src/devices_models/devices/thermal_generation.jl:336]: Do we need to add handling of initial conditions to the common function?

Comment (sourabhdalvi) [reply src/devices_models/devices/thermal_generation.jl:336]: Yes, because most constraints other than range have initial conditions

File: docs/src/man/multistart_formulation.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+	\item[$w \in \cW$] Set of renewable generators.
+	\item[$t \in \cT$] Hourly time steps: $1, \ldots, T$, $T$ = {\tt time\_periods}
+	\item[$l \in \cL_g$] Piecewise production cost intervals for thermal generator $g$: $1, \ldots, L_g$.
+	\item[$s \in \cS_g$] Startup categories for thermal generator $g$, from hottest ($1$) to coldest ($S_g$): $1, \ldots, S_g$.
+\end{itemize}
+```
+
+### System Parameters
+```math
+\begin{itemize}
```
Comment (jd-lara) [inline docs/src/man/multistart_formulation.md:39]: These are pieces of the formulation that pertain to the whole system. We should keep only the ThermalGeneration constraints here and use the references to the other repo if someone is interested in the whole UC.

File: docs/src/man/multistart_formulation.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	\item[$SU_g$]    Startup capability for generator $g$ (MW), {\tt ramp\_startup\_limit}
+	\item[$TS^s_g$] Time offline after which the startup category $s$ becomes active (h), {\tt startup['lag']}.
+	\item[$UT_g$]    Minimum up time for generator $g$ (h), {\tt time\_up\_minimum}.
+	\item[$UT^0_g$] Number of time periods the unit has been on prior to the first time period for generator $g$, {\tt time\_up\_t0}.
+	\item[$U_g^0$]  Initial on/off status for generator $g$, $U_g^0=1$ for $g \in \cG_{\textit{on}}^0$, $U_g^0=0$ for $g \in \cG_{\textit{off}}^0$,  {\tt unit\_on\_t0}.
+	\item[$U_g$] 	Must-run status for generator $g$, {\tt must\_run}.
+\end{itemize}
+```
+
+### Renewable Generator Parameters
```
Comment (jd-lara) [inline docs/src/man/multistart_formulation.md:68]: Same here.

File: src/core/definitions.jl [hunk 7]
```diff
@@ -4,6 +4,7 @@ const MinMax = NamedTuple{(:min, :max), NTuple{2, Float64}}
 const NamedMinMax = Tuple{String, MinMax}
 const UpDown = NamedTuple{(:up, :down), NTuple{2, Float64}}
 const InOut = NamedTuple{(:in, :out), NTuple{2, Float64}}
+const StartUp = NamedTuple{(:hot, :warm, :cold), NTuple{3, Float64}}
```
Comment (jd-lara) [inline src/core/definitions.jl:7]:
```suggestion
const StartUpStages = NamedTuple{(:hot, :warm, :cold), NTuple{3, Float64}}
```

File: src/core/definitions.jl [hunk 8]
```diff
@@ -50,6 +51,7 @@ const MISSING_INITIAL_CONDITIONS_TIME_COUNT = 999.0
 const SECONDS_IN_MINUTE = 60.0
 const MINUTES_IN_HOUR = 60.0
 const SECONDS_IN_HOUR = 3600.0
+const MAX_START_TYPES = 3
```
Comment (jd-lara) [inline src/core/definitions.jl:54]:
```suggestion
const MAX_START_STAGES = 3
```

File: src/core/definitions.jl [hunk 9]
```diff
@@ -109,6 +111,9 @@ const INFLOW = "In"
 const SPILLAGE = "Sp"
 const SLACK_UP = "γ⁺"
 const SLACK_DN = "γ⁻"
+const COLD_START = "δ_cold"
```
Comment (jd-lara) [inline src/core/definitions.jl:114]:
```suggestion
const COLD_START = "start_cold"
```

Keep consistency with other variable names
```

File: src/devices_models/devices/common/cost_functions.jl [hunk 10]
```diff
@@ -182,18 +182,25 @@ function _pwlgencost_sos(
     psi_container::PSIContainer,
     variable::JV,
     cost_component::Vector{NTuple{2, Float64}},
+    status::Union{Nothing, JV} = nothing,
```
Comment (jd-lara) [inline src/devices_models/devices/common/cost_functions.jl:185]: This implementation still troubles me. It needs to be refactored to something more elegant. For now we can leave it.

File: src/devices_models/devices/common/cost_functions.jl [hunk 11]
```diff
@@ -339,6 +347,16 @@ function ps_cost(
     # If array is full of tuples with zeros return 0.0
     all(iszero.(last.(cost_array))) && return JuMP.AffExpr(0.0)
     variable = get_variable(psi_container, var_name)[index, :]
+    settings_ext = get_ext(get_settings(psi_container))
```
Comment (jd-lara) [inline src/devices_models/devices/common/cost_functions.jl:350]: Why is this being passed with ext? There are functions to determine if the psi_container is using parameters.

File: src/devices_models/devices/common/cost_functions.jl [hunk 12]
```diff
@@ -339,6 +347,16 @@ function ps_cost(
...
+    if haskey(settings_ext, "variable_on")
+        var_name = settings_ext["variable_on"]
+        bin = get_variable(psi_container, var_name)[index, :]
+    elseif haskey(settings_ext, "parameter_on")
+        param_name = settings_ext["parameter_on"]
+        bin = get_parameter_container(psi_container, param_name).parameter_array[index, :]
+    else
+        bin = nothing
+    end
```
Comment (jd-lara) [inline src/devices_models/devices/common/cost_functions.jl:359]: I don't follow what you are trying to do here. Can you please elaborate.

File: src/devices_models/devices/common/device_range_constraints.jl [hunk 13]
```diff
@@ -107,23 +113,32 @@ function device_range_constraints!(
...
-            constraint_info = DeviceRangeConstraintInfo(dev_name, limits)
+            if constraint_struct == DeviceRangeConstraintInfo
+                constraint_info = DeviceRangeConstraintInfo(dev_name, limits)
+            elseif constraint_struct == DeviceMultiStartRangeConstraintsInfo
+                lag_limits = rc.lag_limits_func(dev)
+                constraint_info =
+                    DeviceMultiStartRangeConstraintsInfo(dev_name, limits, lag_limits)
+            else
+                error("Missing implementation for $constraint_struct")
+            end
```
Comment (jd-lara) [inline src/devices_models/devices/common/device_range_constraints.jl:141]: looks like this block should be a dispatched function not an if-else conditional.

File: src/devices_models/devices/common/duration_constraints.jl [hunk 14]
```diff
@@ -332,3 +332,99 @@ function device_duration_parameters(
     end
     return
 end
+
+@doc raw"""
+    device_duration_pglib(psi_container::PSIContainer,
```
Comment (jd-lara) [inline src/devices_models/devices/common/duration_constraints.jl:337]: rename to something more descriptive than pglib

File: src/devices_models/devices/common/duration_constraints.jl [hunk 15]
```diff
@@ -332,3 +332,99 @@ function device_duration_parameters(
...
+* psi_container::PSIContainer : the psi_container model built in PowerSimulations
+* duration_data::Vector{UpDown} : gives how many time steps variable needs to be up or down
+* initial_duration::Matrix{InitialCondition} : gives initial conditions for up (column 1) and down (column 2)
+* cons_name::Symbol : name of the constraint
+* var_names::Tuple{Symbol, Symbol, Symbol}) : names of the variables
+- : var_names[1] : varon
+- : var_names[2] : varstart
+- : var_names[3] : varstop
+"""
+function device_duration_pglib(
```
Comment (jd-lara) [inline src/devices_models/devices/common/duration_constraints.jl:370]: Is this code generalizable or is it only applicable to the MultiStart?

File: src/devices_models/devices/common/range_constraint.jl [hunk 16]
```diff
@@ -14,7 +14,7 @@ function RangeConstraintInputsInternal(
         constraint_infos,
         constraint_name,
         variable_name,
-        nothing,
+        Vector{Symbol}(),
```
Comment (jd-lara) [inline src/devices_models/devices/common/range_constraint.jl:17]: Did you make this changes or are they repeat from @daniel-thom changes? it seems odd that this changes are here.

File: src/devices_models/devices/common/range_constraint.jl [hunk 17]
```diff
@@ -221,3 +223,146 @@ function reserve_device_semicontinuousrange(
...
+    varp = get_variable(psi_container, inputs.variable_name)
+    @assert length(inputs.bin_variable_names) == 3
+    varstatus = get_variable(psi_container, inputs.bin_variable_names[1])
+    varon = get_variable(psi_container, inputs.bin_variable_names[2])
+    varoff = get_variable(psi_container, inputs.bin_variable_names[3])
+
+    on_name = middle_rename(inputs.constraint_name, PSI_NAME_DELIMITER, "lb")
+    off_name = middle_rename(inputs.constraint_name, PSI_NAME_DELIMITER, "ub")
+    names = (x.name for x in inputs.constraint_infos)
+    #MOI has a semicontinous set, but after some tests is not clear most MILP solvers support it.
```
Comment (jd-lara) [inline src/devices_models/devices/common/range_constraint.jl:274]: why is this comment here, seems from a copy paste

File: src/devices_models/devices/common/rateofchange_constraints.jl [hunk 18]
```diff
@@ -174,3 +174,119 @@ function device_mixedinteger_rateofchange(
...
+
+    variable = get_variable(psi_container, var_name)
+
+    set_name = (device_name(ic) for ic in initial_conditions[:, 1])
+    con_up = add_cons_container!(psi_container, up_name, set_name, time_steps)
+    con_down = add_cons_container!(psi_container, down_name, set_name, time_steps)
+
+    for (ix, ic) in enumerate(initial_conditions)
+        name = device_name(ic)
+        #constriant (8)
```
Comment (jd-lara) [inline src/devices_models/devices/common/rateofchange_constraints.jl:235]: These comments are references to what?

File: src/devices_models/devices/electric_loads.jl [hunk 19]
```diff
@@ -97,6 +97,7 @@ function make_active_power_constraints_inputs(
                 variable_name = ACTIVE_POWER,
                 limits_func = x -> (min = 0.0, max = PSY.get_activepower(x)),
                 constraint_func = device_range,
+                constraint_struct = DeviceRangeConstraintInfo,
```
Comment (jd-lara) [inline src/devices_models/devices/electric_loads.jl:100]: These changes are also unrelated to this PR. Seems that you need to rebase or merge.

File: src/devices_models/devices/common/range_constraint.jl (hunk 16 above)
Comment (daniel-thom) [reply src/devices_models/devices/common/range_constraint.jl:17]: What you see is the result of collaboration between Sourabh and me to incorporate his new range constraint struct with my refactor effort.

In my implementation there were either 0 or 1 bin variables.  Sourabh needs it to be 0, 1, or 3.  So, we made it an array.  An empty array replaces the 0 case.

File: src/devices_models/devices/electric_loads.jl (hunk 19 above)
Comment (daniel-thom) [reply src/devices_models/devices/electric_loads.jl:100]: The user needs to choose which constraint struct will be built by `device_range_constraints!`.

File: src/devices_models/devices/common/range_constraint.jl (hunk 16 above)
Comment (jd-lara) [reply src/devices_models/devices/common/range_constraint.jl:17]: Ok thanks, it just seemed to me that this PR has too many changes in places where there shouldn't be any

File: src/devices_models/devices/common/device_range_constraints.jl (hunk 13 above)
Comment (daniel-thom) [reply src/devices_models/devices/common/device_range_constraints.jl:141]: That would be better.  But the different constraint structs take different lists of parameters.

**Comment by jd-lara** [comment]
@daniel-thom has clarified the comments related to the code that looked out place. I marked those comments as resolved.

File: src/devices_models/devices/common/cost_functions.jl (hunk 12 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/common/cost_functions.jl:359]: For the SOS2 implementation you need to know if there is variable or a parameter for Status(ON). This depends on the formulation and feedforward. The variable/paramter is used in this constraint ` sum( x_pwl(i) for i in breakpoints) == varon(t)/param_on(t)`. This wasn't needed before as one of the breakpoint was (0,0).

File: src/devices_models/devices/common/cost_functions.jl (hunk 11 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/common/cost_functions.jl:350]: You suggested me to use psi_settings.ext instead of passing kwargs

Comment (jd-lara) [reply src/devices_models/devices/common/cost_functions.jl:350]: But you are using the ext for a field that already exists in the settings....

File: src/devices_models/devices/common/cost_functions.jl (hunk 10 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/common/cost_functions.jl:185]: I don't need that to be optional arg, I'll remove the Union and keep it only JV

File: docs/src/man/multistart_formulation.md (hunk 5 above)
Comment (sourabhdalvi) [reply docs/src/man/multistart_formulation.md:39]: I'll keep the parts related to Thermal UC

File: src/devices_models/devices/common/cost_functions.jl (hunk 10 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/common/cost_functions.jl:185]: I take that back, only case when you don't have status variable is when your using pwl cost for reserves

File: src/devices_models/devices/common/cost_functions.jl (hunk 11 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/common/cost_functions.jl:350]: I'm not using it for that, I need to know if there is a ON variable/parameter in the formulation for that device

**Comment by jd-lara** [comment]
@daniel-thom should we merge this so it can be updated to the newer variable creation as well?

**Comment by daniel-thom** [comment]
> @daniel-thom should we merge this so it can be updated to the newer variable creation as well?

yes
